deepl/api.go

```diff
@@ -0,0 +1,107 @@
+package deepl
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+// Generic API call, takes method, URL parameters and a JSON object to fill,
+// validates & parses the response and unmarshals it into the JSON object,
+// or throws an error.
+// NOTE: Closes the HTTP response that was opened.
+func (c *DeepLClient) apiCall(method string, params url.Values, jsonObject any) error {
+	// If we're debugging, show what was printed out:
+	if c.Debug > 1 {
+		fmt.Fprintf(os.Stderr, "Values being called using %q to API endpoint (%s): %q\n",
+			method,
+			c.Endpoint,
+			params.Encode(),
+		)
+	}
+
+	// http.PostForm() unfortunately doesn't allow us to set headers, and we need to send the authorization
+	// in the headers, not in the body... (gwyneth 20231104)
+	client := &http.Client{}
+	req, err := http.NewRequest(method, c.Endpoint, strings.NewReader(params.Encode()))
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Add("Authorization", "DeepL-Auth-Key " + c.AuthKey)
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if err := validateResponse(resp); err != nil {
+		return err
+	}
+	err = parseResponse(resp.Body, jsonObject)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Validates the response based on its status code, decoding the returned JSON.
+// If the status code is "normal". does nothing (resp remains untouched and open)
+func validateResponse(resp *http.Response) error {
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		// Parsed JSON error data.
+		var data map[string]interface{}
+		baseErrorText := fmt.Sprintf("Invalid response [%d %s]",
+			resp.StatusCode,
+			statusText(resp.StatusCode))
+		// NOTE: code simplification, we now just use the "standard" error codes from `net/http`.
+		// NOTE: on the following code, @Omochice opted for skipping the traditional JSON object struct,
+		// going directly for the semi-raw map[string]interface{} reply instead. (gwyneth 20231103)
+		e := json.NewDecoder(resp.Body).Decode(&data)
+		if e != nil {
+			return fmt.Errorf("%s", baseErrorText)
+		} else {
+			return fmt.Errorf("%s, %s", baseErrorText, data["message"])
+		}
+	}
+	return nil
```

Comment:
@coderabbitai The `validateResponse` function actually does two things: it checks if the status code is outside the 2xx range, _and_ it also tries to validate the JSON code on the `http.Response`. The latter will only be checked _if_ the status code is outside the 2xx range. Technically speaking, this function only validates an _error_ response from the remote server, namely, if such a response is valid JSON or not.
However, your suggestion to add the response body was also included (it doesn't hurt), which, in turn, required some slight code refactoring; which was a good thing, because I could locate a bug and fix it as well!